treematcher.py

```diff
@@ -307,116 +199,150 @@ def smart_lineage(self, constraint):
-    return result
-
-
-def contains(target_node, pattern, name_list):
-    """
-    Shortcut function to find the name(s) within a node and any of it's descendants.
-    :param target_node: Node to be evaluated, given as @.
-    :param pattern: Internal use
-    :param name_list: List of names to be searched for at a node.
-    :return: True if all names in name_list are found, otherwise False.
-    """
-
-    if isinstance(name_list, six.string_types):
-        name_list = [name_list]
-
-    cached_name = pattern.get_cached_attr('name', target_node)
-    result = all(name in cached_name for name in name_list)
-    return result
-
-
-def number_of_species(target_node, pattern, num):
-    """
-    Shortcut function to find the number of species within a node and any of it's descendants.
-    :param target_node: Node to be evaluated, given as @.
-    :param pattern: Internal use.
-    :param num: number of species searched for at a node.
-    :return: True if number of descendant species is equal to num, otherwise False.
-    """
-    cached_species = pattern.get_cached_attr('species', target_node)
-    if num == len(cached_species):
-        result = True
-    else:
-        result = False
-
-    return result
-
-
-def number_of_leaves(target_node, pattern, num):
-    """
-    Shortcut function to find the number of leaves within a node and any of it's descendants.
-    :param target_node: Node to be evaluated, given as @.
-    :param pattern: Internal use.
-    :param num: Number of descendant leaves to be searched for.
-    :return: True if all number of leaves matches num, otherwise False.
-    """
-    cached_name = pattern.get_cached_attr('name', target_node)
-    if num == len(cached_name):
-        result = True
-    else:
-        result = False
-
-    return result
-
-
-def contains_duplication(target_node, pattern):
-    """
-        Shortcut function to find the number of duplication events at or below a node.
-        :param target_node: Node to be evaluated, given as @.
-        :return: True if node is a duplication, otherwise False.
-    """
-    events = pattern.get_cached_attr('evoltype', target_node)
-    return(events.count('D'))
-
-
-def contains_speciation(target_node, pattern):
-    """
-        Shortcut function to find the number of speciation events at or below a node.
-        :param target_node: Node to be evaluated, use @
-        :return: True if node is a duplication, otherwise False.
-    """
-    events = pattern.get_cached_attr('evoltype', target_node)
-    return(events.count('S'))
+        TODO: args doc here...
+        """
+
+        # Creates a local scope containing function names, variables and other
+        # stuff referred within the pattern expressions. We use Syntax() as a
+        # container of those custom functions and shortcuts.
+
+        constraint_scope = {attr_name:getattr(self.syntax, attr_name)
+                            for attr_name in dir(self.syntax)}
```

Comment:
This is the trick for moving all vocabulary to PatternSyntax. All I am doing is to create a dictionary of function names under PAtternSyntax and pointing to the real functions. So function names can be found when evaluating constraints